hcl2template/types.build.go

```diff
@@ -125,3 +132,356 @@ func (p *Parser) decodeBuildConfig(block *hcl.Block) (*BuildBlock, hcl.Diagnosti
+				var diags hcl.Diagnostics
+				if data == nil {
+					diags = append(diags, &hcl.Diagnostic{
+						Summary: fmt.Sprintf("failed loading %s", typeName),
+						Detail:  "unable to prepare provisioner with build variables interpolation",
+					})
+					return nil, diags
+				}
+
+				for _, build := range b.cfg.Builds {
+					for _, from := range build.Sources {
+						src, _ := b.cfg.Sources[from.Ref()]
+
+						variables := make(Variables)
+						for k, v := range data {
+							if value, ok := v.(string); ok {
+								variables[k] = &Variable{
+									DefaultValue: cty.StringVal(value),
+									Type:         cty.String,
+								}
+							}
+						}
+						variablesVal, _ := variables.Values()
+
+						generatedVariables := map[string]cty.Value{
+							sourcesAccessor: cty.ObjectVal(map[string]cty.Value{
+								"type": cty.StringVal(src.Type),
+								"name": cty.StringVal(src.Name),
+							}),
+							buildAccessor: cty.ObjectVal(variablesVal),
+						}
+
+						for _, pb := range build.ProvisionerBlocks {
+							if pb.PType != typeName {
+								continue
+							}
+							return b.cfg.getProvisioner(src, pb, b.cfg.EvalContext(generatedVariables))
+						}
+					}
+				}
+				diags = append(diags, &hcl.Diagnostic{
+					Summary: fmt.Sprintf("failed loading %s", typeName),
+					Detail:  "unable to prepare provisioner with build variables interpolation",
+				})
+				return nil, diags
+			},
+		})
+	}
+
+	if b.CleanupProvisioner.PType != "" {
+		hookedCleanupProvisioner := &packer.HookedProvisioner{
+			b.CleanupProvisioner.Provisioner,
+			b.CleanupProvisioner.config,
+			b.CleanupProvisioner.PType,
+		}
+		hooks[packer.HookCleanupProvision] = []packer.Hook{&packer.ProvisionHook{
+			Provisioners: []*packer.HookedProvisioner{hookedCleanupProvisioner},
+		}}
+	}
+
+	hook := &packer.DispatchHook{Mapping: hooks}
+	// The builder just has a normal Ui, but targeted
+	builderUi := &packer.TargetedUI{
+		Target: b.Name(),
+		Ui:     originalUi,
+	}
+
+	log.Printf("Running builder: %s", b.BuilderType)
+	ts := packer.CheckpointReporter.AddSpan(b.BuilderType, "builder", b.BuilderConfig)
+	builderArtifact, err := b.Builder.Run(ctx, builderUi, hook)
+	ts.End(err)
+	if err != nil {
+		return nil, err
+	}
+
+	artifacts := make([]packer.Artifact, 0, 1)
+
+	// If there was no result, don't worry about running post-processors
+	// because there is nothing they can do, just return.
+	if builderArtifact == nil {
+		return nil, nil
+	}
+
+	errors := make([]error, 0)
+	keepOriginalArtifact := len(b.PostProcessors) == 0
+
+	// This will interpolate build variables by decoding and preparing the post-processor block again
+	generatedData := make(map[string]interface{})
+	artifactSateData := builderArtifact.State("generated_data")
+	if artifactSateData != nil {
+		for k, v := range artifactSateData.(map[interface{}]interface{}) {
+			generatedData[k.(string)] = v
+		}
+	}
+
+	for _, build := range b.cfg.Builds {
```

Comment:
The same happens for the post-processors but here all the post-processors are decoded and prepared again.